### PR TITLE
Add class for guide rate handling for BlackoilWellModel

### PR DESCRIPTION
**Pull request by akva2**
More well code splitting.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp [hunk 1]
```diff
@@ -0,0 +1,588 @@
...
+        this->inj_water = this->inj_water || group.hasInjectionControl(Opm::Phase::WATER);
+        this->inj_gas   = this->inj_gas   || group.hasInjectionControl(Opm::Phase::GAS);
+    }
+}
+
+class GroupTreeWalker
+{
+public:
+    using GroupOp = std::function<void(const Opm::Group&)>;
+    using WellOp = std::function<void(const Opm::Well&)>;
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelGuideRates.cpp:90]: We must `#include <functional>` to use `std:function<>` and `std::reference_wrapper<>`.

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp [hunk 2]
```diff
@@ -0,0 +1,588 @@
...
+    NodeOp preFinish_{nullptr};
+
+    void traverse();
+
+    void startWalk();
+    void discover(std::string_view group);
+    void finish(std::string_view group);
+
+    bool isSeen(std::string_view group) const;
+    std::size_t insertIndex(std::string_view group) const;
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelGuideRates.cpp:141]: We must `#include <cstddef>` in order to use `std::size_t`.  We must also include `<string_view>` to use that type.

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp [hunk 3]
```diff
@@ -0,0 +1,588 @@
...
+    using NodeOp = void (GroupTreeWalker::*)(std::string_view) const;
+
+    std::reference_wrapper<const Opm::Schedule> sched_;
+    int reportStepIdx_;
+
+    GroupOp visitGroup_{};
+    WellOp visitWell_{};
+
+    std::stack<std::string_view, std::vector<std::string_view>> dfsGroupStack_{};
+    std::unordered_set<std::size_t> dfsGroupDiscovered_{};
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelGuideRates.cpp:129]: We must include `<unordered_set>` to use this type.

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+private:
+    using NodeOp = void (GroupTreeWalker::*)(std::string_view) const;
+
+    std::reference_wrapper<const Opm::Schedule> sched_;
+    int reportStepIdx_;
+
+    GroupOp visitGroup_{};
+    WellOp visitWell_{};
+
+    std::stack<std::string_view, std::vector<std::string_view>> dfsGroupStack_{};
```
Comment (bska) [inline opm/simulators/wells/BlackoilWellModelGuideRates.cpp:128]: We must include `<stack>` in order to use this type.

**Review by bska — COMMENTED** [review]
We're missing a number of standard headers here and we should also probably rename the `well_` object to something like `wellModel_` instead, but other than that I'm happy with this change.

File: opm/simulators/wells/BlackoilWellModelGuideRates.hpp [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    void getGuideRateValues(const GuideRate::RateVector& qs,
+                            const bool                   is_inj,
+                            const std::string&           wgname,
+                            data::GuideRateValue&        grval) const;
+
+    data::GuideRateValue getGuideRateValues(const Well& well) const;
+    data::GuideRateValue getGuideRateValues(const Group& group) const;
+    data::GuideRateValue getGuideRateInjectionGroupValues(const Group& group) const;
+
+    const BlackoilWellModelGeneric& well_;
```
[the remark this answers is not in the document]
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGuideRates.hpp]: done

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp (hunk 1 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGuideRates.cpp:100]: done

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp (hunk 2 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGuideRates.cpp:151]: done

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp (hunk 3 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGuideRates.cpp:139]: done

File: opm/simulators/wells/BlackoilWellModelGuideRates.cpp (hunk 4 above)
Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGuideRates.cpp:138]: done

**Comment by akva2** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me know and I'll merge this into the master branch.

**Pull request merged**